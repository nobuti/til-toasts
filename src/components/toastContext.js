import React, { createContext } from 'react'

import useToast from '../hooks/useToast'
const ToastContext = createContext(null)

export const ToastProvider = ({ children }) => {
  const toast = useToast()
  return <ToastContext.Provider value={toast}>{children}</ToastContext.Provider>
}

export default ToastContext
