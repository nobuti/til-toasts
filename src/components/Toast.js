/** @jsx jsx */
import { useEffect, useCallback, useRef } from 'react'
import { jsx, css } from '@emotion/core'

import { useToastContext } from './toastContext'
import Button from './form/Button'

import { ReactComponent as X } from '../assets/x.svg'

export const factory = (type) => {
  const settings = {
    atom: {
      color: '#2548f5',
      title: 'Pet project',
      message: 'Time to code!',
    },
    ghost: {
      color: '#f58325',
      title: 'Terror movie',
      message: "Don't forget the popcorns!",
    },
    virus: {
      color: '#f5eb25',
      title: 'Oh no',
      message: 'Take a break, make some soup and stay safe!',
    },
    lego: {
      color: '#e425f5',
      title: 'Parenting',
      message: 'Time to play!',
    },
    heart: {
      color: '#f52537',
      title: 'Overcoming COVID-19',
      message: 'Distant but together!',
    },
  }

  return settings[type]
}

const TTL = 5 * 1000

const Toast = ({ color, title, message, toast }) => {
  const { removeToast } = useToastContext()
  const remove = useCallback(() => removeToast(toast), [removeToast, toast])

  const eta = useRef(TTL)

  useEffect(() => {
    const timeout = setTimeout(remove, eta.current)
    const created = new Date()

    return () => {
      const now = new Date()
      eta.current = eta.current - (now - created)
      clearTimeout(timeout)
    }
  }, [remove, toast])

  return (
    <div
      css={css`
        border-radius: 4px;
        width: 240px;
        border-left-style: solid;
        border-left-width: 12px;
        border-left-color: ${color};
        padding: 16px;
        background-color: #fff;
        position: relative;
        box-shadow: 0 2px 12px #10248a;
      `}
    >
      <h2
        css={css`
          font-size: 14px;
          font-weight: 500;
          color: #444;
          margin-bottom: ${message ? 4 : 0}px;
        `}
      >
        {title}
      </h2>
      {message && (
        <p
          css={css`
            font-size: 12px;
            font-weight: 400;
            color: #888;
          `}
        >
          {message}
        </p>
      )}

      <Button
        onClick={remove}
        css={css`
          background: transparent;
          padding: 0;
          width: 24px;
          height: 24px;
          color: #ababab;
          position: absolute;
          top: 8px;
          right: 8px;

          svg {
            width: 16px;
            height: 16px;
          }
        `}
      >
        <X />
      </Button>
    </div>
  )
}

export default Toast
