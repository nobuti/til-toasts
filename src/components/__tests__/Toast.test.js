import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { advanceBy, advanceTo, clear } from 'jest-date-mock'

import Toast from '../Toast'
import * as ToastContext from '../toastContext'

describe('Toast', () => {
  const removeToast = jest.fn()
  jest
    .spyOn(ToastContext, 'useToastContext')
    .mockImplementation(() => ({ removeToast }))

  const toast = { color: '#2548f5', title: 'Wadus', message: 'wadus wadus!' }

  beforeEach(() => {
    advanceTo(new Date(2020, 3, 5, 0, 0, 0)) // reset to date time.
  })

  afterEach(clear)

  it('should render properly', () => {
    const { container } = render(<Toast {...toast} />)
    expect(container).toMatchSnapshot()
  })

  it('should hide on click', async () => {
    const { container } = render(<Toast {...toast} />)
    const button = container.querySelector('button')

    fireEvent.click(button, {
      target: button,
    })

    expect(removeToast).toHaveBeenCalled()
  })

  it('should hide after 5 seconds', async () => {
    render(<Toast {...toast} />)

    advanceBy(6000) // advance time 6 seconds

    expect(removeToast).toHaveBeenCalled()
  })
})
