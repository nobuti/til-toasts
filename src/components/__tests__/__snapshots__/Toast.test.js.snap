// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should render properly 1`] = `
<div>
  <div
    class="css-m76tdx-Toast"
  >
    <h2
      class="css-1v9h6x6-Toast"
    >
      Wadus
    </h2>
    <p
      class="css-1jfni4t-Toast"
    >
      wadus wadus!
    </p>
    <button
      class="css-1nw5hej-Toast"
    >
      <svg>
        x.svg
      </svg>
    </button>
  </div>
</div>
`;
